util/fpga/bitstream_bisect.py

```diff
@@ -0,0 +1,406 @@
+                self._git.bisect_bad(commit)
+
+        self._maybe_visualize("Narrowing down ambiguous results")
+        bisect_log = self._git.bisect_run(*slow_command)
+
+        self._maybe_visualize("Final results")
+        self._git.bisect_reset()
+        return bisect_log
+
+    def _maybe_visualize(self, label: str):
+        """Prints a table summarizing the current range of commits."""
+        if not self._visualize:
+            return
+
+        table = rich.table.Table(
+            title="Bitstream Bisect Status: [bold]{}".format(label))
+        table.add_column("Commit")
+        table.add_column("Description")
+        table.add_column("Cached")
+        table.add_column("Judgment")
+
+        bisect_log = self._git.bisect_log()
+        judgment_dict: Dict[CommitHash,
+                            CommitJudgment] = dict(bisect_log.judgments)
+
+        for commit, description in self._git.bisect_view():
+            judgment = judgment_dict.get(commit)
+            table.add_row(commit, rich.markup.escape(description),
+                          "Cached" if commit in self._cache_keys else "",
+                          judgment.name if judgment else "")
+
+        self._console.print(table)
+
+
+app = typer.Typer(pretty_exceptions_enable=False,
+                  rich_markup_mode="rich",
+                  add_completion=False)
+
+
+@app.command(help=__doc__)
+def main(
+    good: str = typer.Option(..., help="Commit-ish for known-good revision."),
+    bad: str = typer.Option(..., help="Commit-ish for known-bad revision."),
+    fast_command: str = typer.Option(
+        ..., help="Command for testing commits with cached bitstreams."),
+    slow_command: Optional[str] = typer.Option(
+        None, help="Command that tests remaining commits in second phase."),
+    git_binary: str = typer.Option("/usr/bin/git", help="Path to Git binary."),
+    verbose: bool = typer.Option(
+        False, help="Log stdout/stderr of every git command."),
+):
+    # Escape the Bazel sandbox. Without this step, git would not find the .git/
+    # directory. Note that this works nicely with git worktrees.
+    workspace_dir = os.environ['BUILD_WORKSPACE_DIRECTORY']
+    os.chdir(workspace_dir)
+
+    # Create a Git controller capable of invoking the system's `git` binary. Use
+    # it to canonicalize potentially-symbolic "commitish" strings so we don't
+    # misinterpret them when a different commit is checked out.
+    git = GitController(git_binary, verbose)
+    good_commit = git.rev_parse(good)
+    bad_commit = git.rev_parse(bad)
+    print("Canonicalized good revision:", good_commit)
+    print("Canonicalized bad revision: ", bad_commit)
+
+    # NOTE: This will generate network traffic!
+    # Fetch a listing of the current bitstream cache.
+    bitstream_cache = BitstreamCache.MakeWithDefaults()
+    bitstream_cache.GetBitstreamsAvailable(refresh=True)
+    bitstream_cache_keys = set(bitstream_cache.available.keys())
```

Comment:
Just a note here that this may not work to bisect across commits affected by the issue that caused https://github.com/lowRISC/opentitan/pull/16631 unless the paging is repeatable.
(I meant to check this out and see if it would work to do that)

It also wont work to bisect with non-default bitstream arguments or stale caches.

Reply:
> It may not work to bisect with non-default bitstream arguments.

Adding the MakeWithDefaults() method does add a bit of maintenance burden. Is there a specific parameter to `BitstreamCache` that you're worried about?

Reply:
Nope, I think the only one that makes sense to even try setting would be --list and that shouldn't have a problem. I just like to think that recording reservations like that might help down the road somehow. (Especially if it were searchable)